### PR TITLE
Vuu 81 definition as json node

### PR DESCRIPTION
**Pull request by vferraro-scottlogic**
### Changes
- use ObjectNode for definition instead of String/JsonNode

### CODE REVIEW

File: layout-server/src/test/java/org/finos/vuu/layoutserver/service/ApplicationLayoutServiceTest.java [hunk 1]
```diff
@@ -75,7 +77,7 @@ public void createApplicationLayout_invalidDefinition_throwsJsonException() {
         String definition = "invalid JSON";
 
         assertThrows(JsonProcessingException.class, () ->
-            service.persistApplicationLayout("user", objectMapper.readTree(definition))
+            service.persistApplicationLayout("user", (ObjectNode) objectMapper.readTree(definition))
```
Comment (pling-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/service/ApplicationLayoutServiceTest.java:80]: We might not need to cast if we do this: `objectMapper.readValue(definition, ObjectNode.class)`. Alternatively, could we make use of `objectNodeConverter`?

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]

File: layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java [hunk 2]
```diff
@@ -1,21 +1,17 @@
...
 import javax.validation.constraints.NotNull;
 
 @Data
 public class LayoutRequestDto {
 
-    /**
-     * The definition of the layout as a string (e.g. stringified JSON structure containing
-     * components)
-     */
     @JsonProperty(value = "definition", required = true)
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java:12]: I think this annotation is used to map JSON properties to the annotated fields. in our case the json properties have the same name as the fields so I think the ObjectMapper should be able to do the mapping without the annotations. I will verify that that's the case and remove the annotations (also for metadata)

File: layout-server/src/main/java/org/finos/vuu/layoutserver/utils/ObjectNodeConverter.java [hunk 3]
```diff
@@ -2,20 +2,22 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import javax.persistence.AttributeConverter;
+import javax.persistence.Converter;
 import java.io.IOException;
 
-public class JsonNodeConverter implements AttributeConverter<JsonNode, String> {
-    private static final Logger logger = LoggerFactory.getLogger(JsonNodeConverter.class);
+@Converter
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/utils/ObjectNodeConverter.java:14]: I don't remember what prompted me to add this annotation here but I think its actually not needed. The [documentation for @Converter](https://docs.oracle.com/javaee%2F7%2Fapi%2F%2F/javax/persistence/Converter.html) says that "A converter class must be annotated with the Converter annotation **or** defined in the object/relational mapping descriptor as a converter." I think we are already doing the latter in Layout and ApplicationLayout models so I think we can get rid of the @Converter annotation here.

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 4]
```diff
@@ -32,14 +34,15 @@
 @ActiveProfiles("test")
 public class LayoutIntegrationTest {
 
-    private static final String DEFAULT_LAYOUT_DEFINITION = "Default layout definition";
+    private static final String LAYOUT_DEFINITION_STRING = "{\"id\":\"main-tabs\"}";
```
Comment (cfisher-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:37]: Why the change from `DEFAULT_LAYOUT_DEFINITION` to `LAYOUT_DEFINITION_STRING`?

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 5]
```diff
@@ -258,7 +264,7 @@ void updateLayout_validIdAndValidRequest_returns204AndLayoutHasChanged() throws
                 initialLayout);
 
         LayoutRequestDto layoutRequest = createValidLayoutRequest();
-        layoutRequest.setDefinition("Updated definition");
+        layoutRequest.setDefinition(objectNodeConverter.convertToEntityAttribute(LAYOUT_DEFINITION_STRING));
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:267]: changed it to use a different definition here

File: layout-server/src/test/java/org/finos/vuu/layoutserver/service/ApplicationLayoutServiceTest.java (hunk 1 above)
Comment (vferraro-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/service/ApplicationLayoutServiceTest.java:80]: also cleaned this up :)

File: layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java:12]: I have removed the annotations

File: layout-server/src/main/java/org/finos/vuu/layoutserver/utils/ObjectNodeConverter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/utils/ObjectNodeConverter.java:14]: I have removed the annotation as it wasn't needed

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 4 above)
Comment (vferraro-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:42]: Oops! not sure how this change came about but it should probably have DEFAULT_* for consistence with the other fields. I will make the change

**Review by cfisher-scottlogic — APPROVED** [review]

File: layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java [hunk 6]
```diff
@@ -1,23 +1,16 @@
...
-     * The definition of the layout as a string (e.g. stringified JSON structure containing
-     * components)
-     */
-    @JsonProperty(value = "definition", required = true)
-    @NotBlank(message = "Definition must not be blank")
-    private String definition;
+    @NotNull(message = "Definition must not be null")
+    private ObjectNode definition;
 
-    @JsonProperty(value = "metadata", required = true)
```
[the remark this answers is not in the document]
Comment (cfisher-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java:20]: Aye, I can't 100% remember why I added it. If the tests pass, it's not needed.

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java (hunk 4 above)
Comment (vferraro-scottlogic) [reply layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java]: I hadn't pushed the change yet as I wanted to fix the conflicts first. It's done now :)

File: layout-server/src/main/java/org/finos/vuu/layoutserver/utils/ObjectNodeConverter.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.persistence.AttributeConverter;
+import java.io.IOException;
+
+public class ObjectNodeConverter implements AttributeConverter<ObjectNode, String> {
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/utils/ObjectNodeConverter.java:12]: Good shout! JsonNodeConverter got revived somehow. I have deleted the file now. Thanks Peter :)

File: layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java [hunk 8]
```diff
@@ -1,23 +1,16 @@
...
-import javax.validation.constraints.NotBlank;
 import javax.validation.constraints.NotNull;
 
 @Data
 public class LayoutRequestDto {
 
-    /**
-     * The definition of the layout as a string (e.g. stringified JSON structure containing
-     * components)
-     */
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/dto/request/LayoutRequestDto.java]: I have added the comment

File: layout-server/src/main/java/org/finos/vuu/layoutserver/dto/response/LayoutResponseDto.java [hunk 9]
```diff
@@ -8,12 +9,7 @@
 public class LayoutResponseDto {
 
     private UUID id;
-
-    /**
-     * The definition of the layout as a string (e.g. stringified JSON structure containing
-     * components)
-     */
-    private String definition;
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply layout-server/src/main/java/org/finos/vuu/layoutserver/dto/response/LayoutResponseDto.java]: done

File: layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java [hunk 10]
```diff
@@ -32,14 +39,16 @@
 @ActiveProfiles("test")
 public class LayoutIntegrationTest {
 
-    private static final String DEFAULT_LAYOUT_DEFINITION = "Default layout definition";
+    private static final String DEFAULT_LAYOUT_DEFINITION_STRING = "{\"id\":\"main-tabs\"}";
+    private static final String DEFAULT_UPDATED_LAYOUT_DEFINITION_STRING = "{\"id\":\"updated-main-tabs\"}";
```
Comment (pling-scottlogic) [inline layout-server/src/test/java/org/finos/vuu/layoutserver/integration/LayoutIntegrationTest.java:43]: Minor semantic thing that's not going to stop me from approving: I would remove the word "default" and just call this `UPDATED_LAYOUT_DEFINITION_STRING`.

**Review by pling-scottlogic — APPROVED** [review]
Good stuff! Just a very minor thing on wording, but that won't stop me from approving.
